cpp/src/join/quadtree_point_in_polygon.cu

```diff
@@ -0,0 +1,573 @@
+                 quad_indices.begin<uint32_t>(),
+                 d_pq_quad_idx.begin());
+
+  // allocate memory to store numbers of points in polygons in all sub-pairs initialized to 0
+  rmm::device_uvector<uint32_t> d_num_hits(num_unit_pairs, stream);
+  thrust::fill(rmm::exec_policy(stream)->on(stream), d_num_hits.begin(), d_num_hits.end(), 0);
+
+  // generate offsets of sub-paris within the orginal pairs
+  thrust::exclusive_scan_by_key(rmm::exec_policy(stream)->on(stream),
+                                d_quad_offset.begin(),
+                                d_quad_offset.end(),
+                                thrust::constant_iterator<int>(1),
+                                d_quad_offset.begin());
+
+  // assemble components in input/output iterators; note d_quad_offset used in both input and output
+  auto quad_block_id_iter = make_zip_iterator(d_pq_quad_idx.begin(), d_quad_offset.begin());
+  auto offset_length_iter = make_zip_iterator(d_quad_offset.begin(), d_quad_length.begin());
+  thrust::transform(rmm::exec_policy(stream)->on(stream),
+                    quad_block_id_iter,
+                    quad_block_id_iter + num_unit_pairs,
+                    offset_length_iter,
+                    [quad_lengths = *d_quad_lengths] __device__(auto const &v) {
+                      uint32_t quad_id    = thrust::get<0>(v);
+                      uint32_t block_id   = thrust::get<1>(v);
+                      uint32_t num_points = quad_lengths.element<uint32_t>(quad_id);
+                      uint32_t offset     = block_id * threads_per_block;
+                      uint32_t length     = num_points < (block_id + 1) * threads_per_block
+                                          ? num_points - block_id * threads_per_block
+                                          : threads_per_block;
+                      return thrust::make_tuple(offset, length);
+                    });
+
+  auto d_quad_offsets  = cudf::column_device_view::create(quad_offsets, stream);
+  auto d_point_indices = cudf::column_device_view::create(point_indices, stream);
+  auto d_point_x       = cudf::column_device_view::create(point_x, stream);
+  auto d_point_y       = cudf::column_device_view::create(point_y, stream);
+  auto d_poly_offsets  = cudf::column_device_view::create(poly_offsets, stream);
+  auto d_ring_offsets  = cudf::column_device_view::create(ring_offsets, stream);
+  auto d_poly_points_x = cudf::column_device_view::create(poly_points_x, stream);
+  auto d_poly_points_y = cudf::column_device_view::create(poly_points_y, stream);
+
+  quad_pip_phase1_kernel<T><<<num_unit_pairs, threads_per_block, 0, stream>>>(d_pq_poly_idx.data(),
+                                                                              d_pq_quad_idx.data(),
+                                                                              d_quad_offset.data(),
+                                                                              d_quad_length.data(),
```

Comment:
Is there a better way to pass a `device_uvector` to a kernel, similar to how we can pass a `cudf::column_device_view`?

Reply:
Nope.